File: src/formats/txt.ts
// Plain text (.txt) adapter — script/transcript extraction.
//
// No timing in the format: on parse, each non-empty line becomes a cue with
// sequential placeholder timing (DEFAULT_DUR each, back-to-back) so the result
// is editable/re-timeable. On export, only the cue text is written (timing is
// dropped) — one cue per block, blank-line separated.

import { newCueId, emptyDocument, type SubtitleDocument } from "../types/subtitle";
import { sortedCues } from "./srt";

const DEFAULT_DUR = 2; // seconds per line when importing untimed text

export function parseTxt(raw: string): SubtitleDocument {
  const doc = emptyDocument("txt");
  const lines = raw.replace(/\r\n/g, "\n").replace(/^﻿/, "").split("\n");
  let t = 0;
  for (const line of lines) {
    const text = line.trim();
    if (!text) continue;
    doc.cues.push({ id: newCueId(), start: t, end: t + DEFAULT_DUR, text });
    t += DEFAULT_DUR;
  }
  return doc;
}

export function serializeTxt(doc: SubtitleDocument): string {
  return sortedCues(doc.cues).map((cue) => cue.text).join("\n\n") + "\n";
}

File: native/Sources/GlyphlineCore/Formats/TXT.swift
// Plain text (.txt) adapter (ported from ../../src/formats/txt.ts).
//
// No timing: on parse, each non-empty line becomes a cue with sequential
// placeholder timing (DEFAULT_DUR each). On export, only cue text is written
// (timing dropped), one cue per block, blank-line separated.

import Foundation

private let TXT_DEFAULT_DUR = 2.0

public func parseTxt(_ raw: String) -> SubtitleDocument {
    var doc = SubtitleDocument.empty(.txt)
    let lines = raw.replacingOccurrences(of: "\r\n", with: "\n").strippingLeadingBOM().components(separatedBy: "\n")
    var t = 0.0
    for line in lines {
        let text = line.trimmed()
        if text.isEmpty { continue }
        doc.cues.append(Cue(id: newCueId(), start: t, end: t + TXT_DEFAULT_DUR, text: text))
        t += TXT_DEFAULT_DUR
    }
    return doc
}

public func serializeTxt(_ doc: SubtitleDocument) -> String {
    sortedCues(doc.cues).map(\.text).joined(separator: "\n\n") + "\n"
}
